src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Unix/PkcsPalOpenSsl.Encrypt.cs

```diff
@@ -0,0 +1,128 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Security.Cryptography;
+using System.Security.Cryptography.Pkcs;
+using System.Security.Cryptography.X509Certificates;
+
+using Microsoft.Win32.SafeHandles;
+
+namespace Internal.Cryptography.Pal.OpenSsl
+{
+    internal sealed partial class PkcsPalOpenSsl : PkcsPal
+    {
+        public sealed override byte[] Encrypt(
+            CmsRecipientCollection recipients,
+            ContentInfo contentInfo,
+            AlgorithmIdentifier contentEncryptionAlgorithm,
+            X509Certificate2Collection originatorCerts,
+            CryptographicAttributeObjectCollection unprotectedAttributes)
+        {
+            int status;
+
+            using (SafeBioHandle contentBio = Interop.Crypto.CreateMemoryBio())
+            using (SafeAsn1ObjectHandle algoOid = Interop.Crypto.ObjTxt2Obj(contentEncryptionAlgorithm.Oid.Value))
+            {
+                Interop.Crypto.CheckValidOpenSslHandle(algoOid);
+                Interop.Crypto.CheckValidOpenSslHandle(contentBio);
+
+                using (SafeCmsHandle cms = Interop.Crypto.CmsInitializeEnvelope(algoOid, out status))
+                {
+                    if (status == 0)
+                        throw new CryptographicException(SR.Cryptography_Cms_UnknownAlgorithm);
+
+                    Interop.Crypto.CheckValidOpenSslHandle(cms);
+
+                    ClassifyAndAddRecipients(cms, recipients);
+
+                    foreach (X509Certificate2 cert in originatorCerts)
+                    {
+                        using (SafeX509Handle certHandle = Interop.Crypto.X509Duplicate(cert.Handle))
+                        {
+                            Interop.Crypto.CheckValidOpenSslHandle(certHandle);
+                            status = Interop.Crypto.CmsAddOriginatorCert(cms, certHandle);
+                            CheckStatus(status);
```

Comment:
What does it mean to add the cert to the cms and then dispose of the cert handle?  Does CmsAddOriginatorCert dup the handle?  And if so, do we need to do it ourselves, or could we just pass cert.Handle in directly?